Build a regression model.

In [34]:
import pandas as pd
import statsmodels.api as sm
import numpy as np
from sklearn.preprocessing import LabelEncoder

In [35]:
df = pd.read_csv("../data/finalDf.csv")
df.drop("Unnamed: 0", axis=1, inplace=True)
df.head()

fsqid                            name   latitude  \
0  2b35d1155b9e408487c4e75d  Sheetla Das Ki Bagiya Shri Ram  23.249586   
1  51336900e4b063789a511ca4                       Pizza inn  23.249586   
2  51c4981d498ed8aad5ce9721          Danish Icecream Palour  23.249586   
3  5128b4fae4b01d5802641e46                    Bake N Shake  23.249586   
4  59f2ef646bd36b5ffe811366                      Kamla Park  23.249586   

   longitude               category  popularity  price  rating stats  \
0  77.395599                Retail,         NaN    NaN     NaN   NaN   
1  77.395599  Fast Food Restaurant,    0.520196    1.0     NaN   NaN   
2  77.395599      Ice Cream Parlor,    0.916330    1.0     NaN   NaN   
3  77.395599           Coffee Shop,    0.764282    1.0     NaN   NaN   
4  77.395599                  Park,    0.931044    NaN     NaN   NaN   

   free_bikes  
0           2  
1           2  
2           2  
3           2  
4           2

In [36]:
df.fillna(0,inplace=True)

In [37]:
df.head()

fsqid                            name   latitude  \
0  2b35d1155b9e408487c4e75d  Sheetla Das Ki Bagiya Shri Ram  23.249586   
1  51336900e4b063789a511ca4                       Pizza inn  23.249586   
2  51c4981d498ed8aad5ce9721          Danish Icecream Palour  23.249586   
3  5128b4fae4b01d5802641e46                    Bake N Shake  23.249586   
4  59f2ef646bd36b5ffe811366                      Kamla Park  23.249586   

   longitude               category  popularity  price  rating stats  \
0  77.395599                Retail,    0.000000    0.0     0.0     0   
1  77.395599  Fast Food Restaurant,    0.520196    1.0     0.0     0   
2  77.395599      Ice Cream Parlor,    0.916330    1.0     0.0     0   
3  77.395599           Coffee Shop,    0.764282    1.0     0.0     0   
4  77.395599                  Park,    0.931044    0.0     0.0     0   

   free_bikes  
0           2  
1           2  
2           2  
3           2  
4           2

In [39]:
labelencoder = LabelEncoder()
df['category_new'] = labelencoder.fit_transform(df['category'].astype(str))
df.head(5)

fsqid                            name   latitude  \
0  2b35d1155b9e408487c4e75d  Sheetla Das Ki Bagiya Shri Ram  23.249586   
1  51336900e4b063789a511ca4                       Pizza inn  23.249586   
2  51c4981d498ed8aad5ce9721          Danish Icecream Palour  23.249586   
3  5128b4fae4b01d5802641e46                    Bake N Shake  23.249586   
4  59f2ef646bd36b5ffe811366                      Kamla Park  23.249586   

   longitude               category  popularity  price  rating stats  \
0  77.395599                Retail,    0.000000    0.0     0.0     0   
1  77.395599  Fast Food Restaurant,    0.520196    1.0     0.0     0   
2  77.395599      Ice Cream Parlor,    0.916330    1.0     0.0     0   
3  77.395599           Coffee Shop,    0.764282    1.0     0.0     0   
4  77.395599                  Park,    0.931044    0.0     0.0     0   

   free_bikes  category_new  
0           2           111  
1           2            58  
2           2            79  
3           2            38  
4           2           100

In [40]:
# Prepare the inputs for the regression model
x = df[['latitude', 'longitude','popularity','price','rating','category_new']]
x = sm.add_constant(x) # Add a constant term to the inputs
y = df['free_bikes']

In [41]:
# Fit the regression model
model = sm.OLS(y, x).fit()

Provide model output and an interpretation of the results. 

In [43]:
# Print the summary
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:             free_bikes   R-squared:                       0.222
Model:                            OLS   Adj. R-squared:                  0.217
Method:                 Least Squares   F-statistic:                     44.79
Date:                Mon, 30 Jan 2023   Prob (F-statistic):           2.55e-48
Time:                        19:45:24   Log-Likelihood:                -1558.8
No. Observations:                 949   AIC:                             3132.
Df Residuals:                     942   BIC:                             3166.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const         -619.2951    124.278     -4.983   

Model Interpretation:
A measure of the extent to which changes in our independent variables influence our dependent variable is called R-squared. We have a good fit to the data when our R squared is higher. The Adjusted R-squared should be used instead of the R-squared for multivariate linear regression. As adjusted R-squared adjusts the formula based on the number of variables in your model, a lower adjusted score may indicate that some variables aren't contributing adequately to your model's R-squared. It is found that this model has an Adjusted R-squared of 0.217.

The p-value column in the model output shows the effect of each variable on the outcome. If a variable has a p-value greater than the threshold value of 0.05, this may indicate that the variable should be removed from the model in order to get a better model fit. The p-value of rating is found to be significant below 0.05. Hence, the alternative hypothesis is accepted between rating and free_bikes.

# Stretch

How can you turn the regression model into a classification model?

A regression model can be turned in to a classification model by converting the continuous output of the regression model into a categorical output. The output of the regression model can be transformed into class labels by assigning a class label to the output value based on its relation to the threshold.